sys/senml/senml.c

```diff
@@ -0,0 +1,357 @@
+
+            insert_pos += chars_written;
+            cnt_len -= chars_written;
+        }
+
+        if (curr_record->unit) {
+            chars_written = snprintf(&output[insert_pos], cnt_len, "\"u\":\"%s\",",
+                                     curr_record->unit);
+
+            if (chars_written >= cnt_len || chars_written < 0) {
+                return -1;
+            }
+
+            insert_pos += chars_written;
+            cnt_len -= chars_written;
+        }
+
+        if (curr_record->time != 0) {
+            chars_written = snprintf(&output[insert_pos], cnt_len, "\"t\":%f,",
+                                     curr_record->time);
+
+            if (chars_written >= cnt_len || chars_written < 0) {
+                return -1;
+            }
+
+            insert_pos += chars_written;
+            cnt_len -= chars_written;
+        }
+
+        if (curr_record->update_time != 0) {
+            chars_written = snprintf(&output[insert_pos], cnt_len, "\"ut\":%f,",
+                                     curr_record->update_time);
+
+            if (chars_written >= cnt_len || chars_written < 0) {
+                return -1;
+            }
+
+            insert_pos += chars_written;
+            cnt_len -= chars_written;
+        }
+
+        if (curr_record->value_type != SENML_TYPE_UNDEF) {
+            if (curr_record->value_type == SENML_TYPE_FLOAT) {
+                chars_written = snprintf(&output[insert_pos], cnt_len, "\"v\":%f,",
+                                         curr_record->value.f);
+            }
+            else if (curr_record->value_type == SENML_TYPE_STRING) {
+                chars_written = snprintf(&output[insert_pos], cnt_len, "\"vs\":\"%s\",",
+                                         curr_record->value.s);
+            }
+            else if (curr_record->value_type == SENML_TYPE_BOOL) {
+                chars_written = snprintf(&output[insert_pos], cnt_len, "\"vb\":%s,",
+                                         curr_record->value.b ? "true" : "false");
+            }
+            else if (curr_record->value_type == SENML_TYPE_BINARY) {
+                chars_written = snprintf(&output[insert_pos], cnt_len, "\"vd\":\"%s\",",
+                                         curr_record->value.d);
+            }
+
+            if (chars_written >= cnt_len || chars_written < 0) {
+                return -1;
+            }
+
+            insert_pos += chars_written;
+            cnt_len -= chars_written;
+        }
+
+        // decrease insert_pos to overwrite trailing comma
+        insert_pos--;
+        cnt_len++;
+
+        output[insert_pos] = '}';
+        insert_pos++;
+        cnt_len--;
+
+        output[insert_pos] = ',';
+        insert_pos++;
+        cnt_len--;
+    }
+
+    // remove trailing comma or unnecessary opening curly brace
+    if (insert_pos >= 2) {
+        insert_pos--;
+    }
+
+    output[insert_pos] = ']';
+    insert_pos++;
+
+    output[insert_pos] = '\0';
+    insert_pos++;
```

Comment:
`ssize_t cnt_len = len - 3;` can be `ssize_t cnt_len = len - 1;`.

Invariant:
`insert_pos + cnt_len == len - 3`
so
`insert_pos == len - 3 - cnt_len`.

Abstract execution trace:

```C
    // ...

    output[insert_pos] = ',';
    insert_pos++;
    cnt_len--;
    ///////////////////////////
    // cnt_len >= 0          //
    // insert_pos <= len - 3 //
    ///////////////////////////
}

// remove trailing comma or unnecessary opening curly brace
if (insert_pos >= 2) {
    insert_pos--;
    ///////////////////////////   
    // insert_pos <= len - 4 //
    ///////////////////////////
}

output[insert_pos] = ']';
insert_pos++;
///////////////////////////   
// insert_pos <= len - 3 //
///////////////////////////

output[insert_pos] = '\0';
insert_pos++;
///////////////////////////   
// insert_pos <= len - 2 //
///////////////////////////
```

`insert_pos` should be `<= len` here instead of `<= len - 2`, so initial value of `cnt_len` should be `len - 1` rather than `len - 3`.